### PR TITLE
Add FreeBSD pkg support

### LINKED ISSUE #21: Implement helper actions for FreeBSD pkg
The module provide helpers (e.g. for listing outdated packages) which use custom code since the Puppet providers are not providing this information.

Implementing these helpers for pkg would allow to use these helpers on FreeBSD.

### LINKED ISSUE #23: Add a refresh actions to update available packages
Some package managers provide different actions for refreshing available package list and for listing available updates.  apt is one of them and therefore has 2 actions:
  - apt_update
  - apt_checkupdates

The `checkupdates` actions call `apt_checkupdates` when the package provider is apt, but `update` is _not_ an alias for `apt_update`: it is used to actually update a package to the latest available version.

With a single `refresh` action, we can act regardless of the package manager.  This action should update the package list if available and return the list of available updates as checkupdate does.

### PR DESCRIPTION
**Pull request by smortex**
```sh-session
romain@zappy ~ $ mco package checkupdates -I `hostname -f`

 * [ ==========================================================> ] 1 / 1

   zappy.home.sigabrt.org: audacity c-ares check cups curl dav1d dbus devcpu-data e2fsprogs e2fsprogs-libblkid ffmpeg firefox glib gtk-update-icon-cache gtk3 harfbuzz harfbuzz-icu lcms2 libev mencoder mesa-dri mesa-libs meson mplayer mumble mutt nspr nss p5-CGI p5-MIME-Base32 pdfcrack pdftk pkg pkgconf py37-more-itertools rubygem-ast rubygem-aws-partitions rubygem-aws-sdk-core rubygem-aws-sdk-ec2 rubygem-aws-sigv4 rubygem-bolt rubygem-parallel rubygem-parser rubygem-r10k sudo youtube_dl


Finished processing 1 / 1 hosts in 9520.61 ms

romain@zappy ~ $ mco package count -I `hostname -f`

 * [ ==========================================================> ] 1 / 1

   zappy.home.sigabrt.org: 899

Summary of Count:

   899 = 1


Finished processing 1 / 1 hosts in 1141.52 ms

romain@marvin ~ $
```

Fixes #21 
Fixes #23

### CODE REVIEW

**Comment by smortex** [comment]
The `pkg_checkupdates` implementation does not refresh the package lists (just like the `apt_checkupdates` action).  In order to refresh the package list on debian, the module has a `apt_update` action, so we can add a `pkg_update` action.  My concern is about one action per provider which is cumbersome.

The module provide a `checkupdate` action that will run `apt_chekcupdates`, `yum_checkupdates` (and now `pkg_checkupdates`) depending on what package management is available.  A similar single entry point for updating the package lists might be desirable.

If so, such method not existing yet, and `update` being already used to update a single package, how to name it?  `refresh`?

**Comment by ripienaar** [comment]
refresh makes sense yes, be good to have such a action

**Comment by smortex** [comment]
Done.  I removed the WIP flag

File: util/package/packagehelpers.rb [hunk 1]
```diff
@@ -215,6 +265,68 @@ def self.apt_checkupdates(output = "")
...
+        end
+
+        def self.refresh
+          manager = packagemanager
+          if manager == :apt
+            return apt_update
+          elsif manager == :pkg
+            return pkg_update
+          elsif manager == :yum
+            return yum_checkupdates
```
Comment (ripienaar) [inline util/package/packagehelpers.rb:323]: problem here is that `yum_checkupates` does not imply a metadata refresh, yum has a policy for how oftern refreshes happen, we can maybe call `yum_clean("metadata")` before that would force it. Not sure how zypper works unfortunately

File: agent/package.ddl [hunk 2]
```diff
@@ -248,3 +248,17 @@ action "checkupdates", :description => "Check for updates" do
...
 end
+
+action "refresh", :description => "Update the available packages cache" do
+    output :output,
+           :description => "Output from the package manager",
+           :display_as  => "Output"
+
+    output :outdated_packages,
+           :description => "Outdated packages",
+           :display_as  => "Outdated Packages"
```
Comment (ripienaar) [inline agent/package.ddl:259]: wonder if refresh really should imply check outdated?

File: util/package/packagehelpers.rb [hunk 3]
```diff
@@ -215,6 +265,68 @@ def self.apt_checkupdates(output = "")
...
+          elsif manager == :zypper
+            return zypper_checkupdates
+          else
+            raise 'Cannot find a compatible package system to update packages'
```
[the remark this answers is not in the document]
Comment (smortex) [reply util/package/packagehelpers.rb:327]: Oops!  Might be cool to update `.rubocop.yml` (it fails with recent rubocop) and add a CI test to enforce this.

**Comment by ripienaar** [comment]
Yeah I want to do rubocop rules copier from the mcollective-choria repo. But that requires time to fix up the nightmare that will cause foo :) so plugin repos is a bit of Wild West style wise till I or someone get time/inclination

File: agent/package.ddl (hunk 2 above)
Comment (smortex) [reply agent/package.ddl:259]: It's what `apt_update` did, so I went the same way.  In my use case, I ran `apt_update` once and then `checkupdates` as many times as needed while performing updates.

I was thinking replacing `apt_update` with `refresh` and keeping that behavior, but it's really not needed.  Do you prefer that the rpc only returns an error code and the output from the command and that the application list failed nodes?

File: util/package/packagehelpers.rb (hunk 1 above)
Comment (smortex) [reply util/package/packagehelpers.rb:323]: Good catch, I start to understand why there was no `yum_update` and no `apt_clean`… I was thinking that yum would auto-refresh repositories when needed but it seems to be more complicated…

Regarding zypper, [this page](http://linuxg.net/how-to-install-find-and-remove-packages-with-zypper/) suggest that `zypper refresh` should be used to update the repo list.

Let's fix this.

**Comment by smortex** [comment]
I pushed commits implementing the changes / suggestions above.  Depending on your though on this, I will either squash them in the appropriate commit or drop them.  Thanks!

**Comment by ripienaar** [comment]
looking good, we got some failing tests - and squash commits then its good to go :)

**Comment by smortex** [comment]
Thanks for the feedback!
Test suite adjusted and changes squashed in 2 commits, one per issue fixed.

**Pull request merged**